oximeter/collector/src/self_stats.rs

```diff
@@ -0,0 +1,154 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Metrics oximeter reports about itself
+
+// Copyright 2023 Oxide Computer Company
+
+use crate::ProducerEndpoint;
+use oximeter::types::Cumulative;
+use oximeter::types::ProducerResultsItem;
+use oximeter::Metric;
+use oximeter::MetricsError;
+use oximeter::Sample;
+use oximeter::Target;
+use std::collections::BTreeMap;
+use std::net::IpAddr;
+use std::time::Duration;
+use uuid::Uuid;
+
+/// The interval on which we report self statistics
+pub const COLLECTION_INTERVAL: Duration = Duration::from_secs(60);
+
+/// A target representing a single oximeter collector.
+#[derive(Clone, Copy, Debug, Target)]
+pub struct OximeterCollector {
+    /// The collector's ID.
+    pub collector_id: Uuid,
+    /// The collector server's IP address.
+    pub collector_ip: IpAddr,
+    /// The collector server's port.
+    pub collector_port: u16,
+}
+
+/// The number of successful collections from a single producer.
+#[derive(Debug, Metric)]
+pub struct Collections {
+    /// The producer's ID.
+    pub producer_id: Uuid,
+    /// The producer's IP address.
+    pub producer_ip: IpAddr,
+    /// The producer's port.
+    pub producer_port: u16,
+    /// The base route in the producer server used to collect metrics.
+    ///
+    /// The full route is `{base_route}/{producer_id}`.
+    pub base_route: String,
+    pub datum: Cumulative<u64>,
+}
+
+/// Small enum to help understand why oximeter failed to collect from a
+/// producer.
+#[derive(
+    Clone,
+    Copy,
+    Debug,
+    Eq,
+    Hash,
+    Ord,
+    PartialEq,
+    PartialOrd,
+    strum::EnumIter,
+    strum::Display,
+)]
+#[non_exhaustive]
+#[strum(serialize_all = "snake_case")]
+pub enum FailureReason {
+    /// The producer could not be reached.
+    Unreachable,
+    /// Error during deserialization.
+    Deserialization,
+    /// Some unknown reason.
+    Other,
+}
+
+/// The number of failed collections from a single producer.
+#[derive(Debug, Metric)]
+pub struct FailedCollections {
+    /// The producer's ID.
+    pub producer_id: Uuid,
+    /// The producer's IP address.
+    pub producer_ip: IpAddr,
+    /// The producer's port.
+    pub producer_port: u16,
+    /// The base route in the producer server used to collect metrics.
+    ///
+    /// The full route is `{base_route}/{producer_id}`.
+    pub base_route: String,
+    /// The reason we could not collect.
+    //
+    // NOTE: This should always be generated through a `FailureReason`.
+    pub reason: String,
```

Comment:
No, it can't. We don't support arbitrary _typed_ values in oximeter, a field needs to be a variant of [`FieldType`](https://github.com/oxidecomputer/omicron/blob/85ac741275648bf21517a91a4d155977c425cb79/oximeter/oximeter/src/types.rs#L48), hence the conversion to a string here.